### PR TITLE
`check_new_syntax.py`: check for any import of deprecated `contextlib` aliases

### PR DESCRIPTION
**Pull request by AlexWaygood**
For most deprecated aliases in the `typing` module, we're not yet able to test for their usage in class definitions in `check_new_syntax.py`, due to `__class_getitem__` not being available in earlier versions of Python. However, with `contextlib`, we're able to freely use the new-style syntax without causing errors on earlier versions of Python. Therefore, we should test for presence of the deprecated aliases in `check_new_syntax.py`.

My plan for this PR is:

1. Run CI on it once, to check that code using pattern-matching doesn't break the CI (if it does, convert it to code using `if-elif` and `isinstance()`, etc.)
2. Deliberately restore some deprecated syntax to demonstrate that the test fails, run the CI again.
3. Revert the deliberate reintroduction of deprecated syntax, making this PR mergeable.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
@JelleZijlstra, what magic words would I have to add to the `pre-commit-config.yaml` file to stop it erroring out? 🙂 You need the `--target-version py310` command-line argument with black if you want it to be able to handle pattern-matching (https://github.com/psf/black/blob/main/CHANGES.md) -- where do I add that in the `pre-commit-config.yaml` file?

**Comment by Akuli** [comment]
Try changing `target_version = ["py37"]` in `pyproject.toml`.

**Comment by AlexWaygood** [comment]
> My plan for this PR is:
> 
> 1. Run CI on it once, to check that code using pattern-matching doesn't break the CI (if it does, convert it to code using `if-elif` and `isinstance()`, etc.)
> 2. Deliberately restore some deprecated syntax to demonstrate that the test fails, run the CI again.
> 3. Revert the deliberate reintroduction of deprecated syntax, making this PR mergeable.

Stage 1: complete. (Thanks @Akuli!)

**Comment by AlexWaygood** [comment]
The last commit caused more errors than I intended, as there's now an unused import in `mmap.pyi` that's causing pyright to complain. But, you can see that `check_new_syntax.py` is raising errors in the intended way w.r.t. the reintroduction of deprecated syntax in `mmap.pyi`.

**Comment by AlexWaygood** [comment]
> My plan for this PR is:
> 
> 1. Run CI on it once, to check that code using pattern-matching doesn't break the CI (if it does, convert it to code using `if-elif` and `isinstance()`, etc.)
> 2. Deliberately restore some deprecated syntax to demonstrate that the test fails, run the CI again.
> 3. Revert the deliberate reintroduction of deprecated syntax, making this PR mergeable.

The last commit caused more errors than I intended, as there's now an unused import in `mmap.pyi` that's causing pyright to complain. But, you can see that `check_new_syntax.py` is raising errors in the intended way w.r.t. the reintroduction of deprecated syntax in `mmap.pyi`.

Stage 2: complete.

**Comment by AlexWaygood** [comment]
> My plan for this PR is:
> 
> 1. Run CI on it once, to check that code using pattern-matching doesn't break the CI (if it does, convert it to code using `if-elif` and `isinstance()`, etc.)
> 2. Deliberately restore some deprecated syntax to demonstrate that the test fails, run the CI again.
> 3. Revert the deliberate reintroduction of deprecated syntax, making this PR mergeable.

Stage 3: complete.

File: .github/workflows/tests.yml
Comment (Akuli) [inline .github/workflows/tests.yml:47]: Python 3.10 for black is now specified in 5 different places. Do we need it anywhere else than in pyproject.toml?

File: tests/check_new_syntax.py
Comment (Akuli) [inline tests/check_new_syntax.py:51]: We can get rid of the `e.g. foo` part to simplify this a bit. The error is clear enough without it.

File: tests/check_new_syntax.py
Comment (Akuli) [inline tests/check_new_syntax.py:44]: The function naming is a bit confusing, because the functions named `raise_blah()` don't actually raise an error, but append it to a list.

File: tests/check_new_syntax.py
Comment (Akuli) [inline tests/check_new_syntax.py:168]: Maybe it would make sense to run  `OldSyntaxFinder()` on the base classes? Maybe add an argument specifying whether it is for a base class or not, so that it doesn't complain about stuff that doesn't apply to base classes? Checking base classes too would also fix this comment `# TODO: can use built-in generics in type aliases`. If we do that, we probably don't need the `target_version` stuff anymore.

File: .github/workflows/tests.yml
Comment (AlexWaygood) [reply .github/workflows/tests.yml:47]: No idea! Let's find out!

Comment (AlexWaygood) [reply .github/workflows/tests.yml]: Looks like we don't!

File: tests/check_new_syntax.py
Comment (AlexWaygood) [reply tests/check_new_syntax.py]: That was my initial thought, as well, but I think it makes sense to keep the code separate. There are actually several things different about the code that needs to run on class definitions compared to the code that needs to run on annotations (though the code that needs to run on class definitions can be combined easily with the code that needs to run on `TypeVar` declarations -- my planned next PR 🙂.) So, I think I'd rather not use `OldSyntaxFinder()` here.

I am, however, happy to change this code block from using `match`/`case` syntax to using more backwards-compatible syntax, if you'd rather not bump `pyproject.toml` from py37 to py310. I personally find the `match`/`case` syntax much more readable and elegant in this case, but I can see why you might not want to bump the `target_version` in `pyproject.toml`.

Comment (AlexWaygood) [reply tests/check_new_syntax.py]: With regard to type aliases: yeah, I don't really want to touch them right now. The idea of figuring out which aliases are okay and which aren't looks kinda complicated and horrible.

**Comment by AlexWaygood** [comment]
(I've done manual testing after the last two commits, and the test still operates as intended.)

**Comment by Akuli** [comment]
I'll let other maintainers decide whether we even want to check for contextlib aliases. Originally `check_new_syntax.py` was intended to tell new contributors to use `list` instead of `List`, `|` instead of `Union` and the like. After thinking about this more, I don't think we really need to check for things that come up less often.

**Comment by AlexWaygood** [comment]
> After thinking about this more, I don't think we really need to check for things that come up less often.

I see your point. On the other hand, these aliases are deprecated, right? Meaning they'll presumably be removed from Python at some point? Checking for them in this test could save us having to make changes in the future when the aliases are eventually removed from the `typing` module.

**Comment by Akuli** [comment]
>Checking for them in this test could save us having to make changes in the future when the aliases are eventually removed from the typing module.

Honestly this doesn't seem like something that `git grep` wouldn't do very conveniently. The aliases even have different names.

**Comment by AlexWaygood** [comment]
> Honestly this doesn't seem like something that `git grep` wouldn't do very conveniently. The aliases even have different names.

Fair!

**Comment by Akuli** [comment]
Maybe we should check them in imports? If you import a deprecated alias, it should be changed IMO. This is different from `Optional` and `Union`, because they still have use in base classes and type aliases.

**Comment by AlexWaygood** [comment]
> Maybe we should check them in imports? If you import a deprecated alias, it should be changed IMO. This is different from `Optional` and `Union`, because they still have use in base classes and type aliases.

Yes, that seems like a better approach!

**Comment by AlexWaygood** [comment]
The changes I've just pushed are essentially a complete rewrite of all `contextlib`-related checks in `check_new_syntax.py`. I've once again deliberately reintroduced some deprecated syntax to demonstrate that the test works; I'll revert that as soon as the CI run is done.

**Review by Akuli — APPROVED** [review]
Thanks! I like it when the code shortens and at the same time gets a new feature implemented :)

**Comment by AlexWaygood** [comment]
> Thanks! I like it when the code shortens and at the same time gets a new feature implemented :)

Don't we all 😃

**Pull request merged**
